training/CUQIpy-PyTorch/HMC.ipynb

```diff
@@ -0,0 +1,857 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Because the `cuqi` distributions are written using NumPy and SciPy, we instead have to use the distributions defined in `cuqipy_pytorch`.\n",
+    "\n",
+    "*We are thinking of making the main CUQIpy package agnostic to the backend such that it can use either NumPy or PyTorch (or Jax, or TensorFlow etc.), but this is not yet implemented.*\n",
+    "\n",
+    "Instead, we use thin wrappers around PyTorch distributions, that acts as a drop-in replacement for the `cuqi` distributions. They are imported as follows."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from cuqipy_pytorch.distribution import Gaussian, HalfGaussian, LogGaussian, Uniform, Gamma, StackedJointDistribution"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "CUQIpy-PyTorch distributions work in a similar way to the `cuqi` distributions, but instead of using NumPy arrays, they use PyTorch tensors. For example, we can create a 2D i.i.d. Gaussian distribution and look at some of its properties and methods as follows."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "x = Gaussian(xp.zeros(2), xp.eye(2))\n",
+    "\n",
+    "print(f\" Name: {x.name}\")\n",
+    "print(f\" Mean: {x.mean}\")\n",
+    "print(f\" Cov: {x.cov}\")\n",
+    "print(f\" PDF at 0: {x.pdf(xp.zeros(2))}\")\n",
+    "print(f\" Log PDF at 0: {x.logpdf(xp.zeros(2))}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The major difference is that the gradient of the logpdf is automatically computed using PyTorch's automatic differentiation feature. We can inspect the code for the `gradient` method to see how this is done. This should look similar to the example we did for PyTorch gradients above."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Evaluate gradient for some random point\n",
+    "x.gradient(xp.randn(2))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Inspect the code for evaluating gradient\n",
+    "x.gradient??"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Really utilizing the power of automatic differentiation\n",
+    "\n",
+    "One of the main use-cases for automatic differentiation in the context of Bayesian inference is the compute the gradient of more complex Bayesian models. To illustrate this, suppose one part of the Bayesian model contains a Gaussian distribution with a covariance matrix that is a function of another random variable, say $s$. The Gaussian could be written as follows.\n",
+    "\n",
+    "$$\n",
```

Comment:
This math is unrendered on colab
![image](https://user-images.githubusercontent.com/9976072/192164128-dc7d999f-cf26-4468-82a3-6260ef25a27b.png)

Reply:
Thanks